torch/utils/tensorboard/summary.py

```diff
@@ -0,0 +1,462 @@
+
+
+def image_boxes(tag, tensor_image, tensor_boxes, rescale=1, dataformats='CHW'):
+    '''Outputs a `Summary` protocol buffer with images.'''
+    tensor_image = make_np(tensor_image)
+    tensor_image = convert_to_HWC(tensor_image, dataformats)
+    tensor_boxes = make_np(tensor_boxes)
+    tensor_image = tensor_image.astype(
+        np.float32) * _calc_scale_factor(tensor_image)
+    image = make_image(tensor_image.astype(np.uint8),
+                       rescale=rescale,
+                       rois=tensor_boxes)
+    return Summary(value=[Summary.Value(tag=tag, image=image)])
+
+
+def draw_boxes(disp_image, boxes):
+    # xyxy format
+    num_boxes = boxes.shape[0]
+    list_gt = range(num_boxes)
+    for i in list_gt:
+        disp_image = _draw_single_box(disp_image,
+                                      boxes[i, 0],
+                                      boxes[i, 1],
+                                      boxes[i, 2],
+                                      boxes[i, 3],
+                                      display_str=None,
+                                      color='Red')
+    return disp_image
+
+
+def make_image(tensor, rescale=1, rois=None):
+    """Convert an numpy representation image to Image protobuf"""
+    from PIL import Image
+    height, width, channel = tensor.shape
+    scaled_height = int(height * rescale)
+    scaled_width = int(width * rescale)
+    image = Image.fromarray(tensor)
+    if rois is not None:
+        image = draw_boxes(image, rois)
+    image = image.resize((scaled_width, scaled_height), Image.ANTIALIAS)
+    import io
+    output = io.BytesIO()
+    image.save(output, format='PNG')
+    image_string = output.getvalue()
+    output.close()
+    return Summary.Image(height=height,
+                         width=width,
+                         colorspace=channel,
+                         encoded_image_string=image_string)
+
+
+def video(tag, tensor, fps=4):
+    tag = _clean_tag(tag)
+    tensor = make_np(tensor)
+    tensor = _prepare_video(tensor)
+    # If user passes in uint8, then we don't need to rescale by 255
+    scale_factor = _calc_scale_factor(tensor)
+    tensor = tensor.astype(np.float32)
+    tensor = (tensor * scale_factor).astype(np.uint8)
+    video = make_video(tensor, fps)
+    return Summary(value=[Summary.Value(tag=tag, image=video)])
+
+
+def make_video(tensor, fps):
+    try:
+        import moviepy  # noqa: F401
+    except ImportError:
+        print('add_video needs package moviepy')
+        return
+    try:
+        from moviepy import editor as mpy
+    except ImportError:
+        print("moviepy is installed, but can't import moviepy.editor.",
+              "Some packages could be missing [imageio, requests]")
+        return
+    import tempfile
+
+    t, h, w, c = tensor.shape
+
+    # encode sequence of images into gif string
+    clip = mpy.ImageSequenceClip(list(tensor), fps=fps)
+    with tempfile.NamedTemporaryFile() as f:
```

Comment:
I don't understand how this could possibly work. By default NamedTemporaryFile deletes itself on closure. Use of a file object as context manager is supposed to close it when it exits scope. But you use the file object afterwards. Very fishy!

Reply:
Figured it out - we are actually only using the *name* of the NamedTemporaryFile() and not the file itself. We should really change this code, though. Tracked with a task.